### ISSUE TITLE
Balena push does not build in local mode when compose file specifies a build context

### ISSUE THREAD

**Issue by joelmarty** [balena-io/balena-cli#2715]
I created a project using a docker-compose file to deploy on a raspberry pi zero w device.
The compose file specifies one service which is build from a subdirectory in my repo.
When switching the device to local mode and deploying using `balena push`, the command hangs indefinitely and the image is never built.

The folder structure is as follows:
```
repo
|_ pihole
  |_ Dockerfile
  |_ ...
|_ docker-compose.yml
```

# Expected Behavior

`balena push <my device ip>` builds the project

# Actual Behavior

`balena -push x.x.x.x` hangs.
Here is a log trace with debug logging enabled:
```
$ balena --debug push x.x.x.x 
[debug] new argv=[/home/joel/opt/balena-cli/balena,/snapshot/balena-cli/bin/balena,push,x.x.x.x] length=4
[debug] Deprecation check: 3.06398 days since last npm registry query for next major version release date.
[debug] Will not query the registry again until at least 7 days have passed.
[Debug]   Using build source directory: . 
[Debug]   Pushing to local device: x.x.x.x
[Debug]   Checking we can access device
[Debug]   Sending request to http://x.x.x.x:48484/ping
[debug] Event tracking error: Timeout awaiting 'response' for 0ms
[Debug]   Checking device supervisor version: 15.0.4
[Info]    Starting build on device x.x.x.x
[Debug]   Loading project...
[Debug]   Resolving project...
[Debug]   docker-compose.yml file found at "."
[Debug]   Creating project...
[Debug]   Tarring all non-ignored files...
[Debug]   Tarring complete in 1090 ms
[Debug]   Fetching device information...
[Debug]   Sending request to http://x.x.x.x:48484/v2/local/device-info
[Debug]   Found build tasks:
[Debug]       pihole: build [./pihole]
[Debug]   Resolving services with [raspberry-pi|rpi]
[Debug]   Found project types:
[Debug]       pihole: Standard Dockerfile
[Debug]   Probing remote daemon for cache images
[Debug]   Using 3 on-device images for cache...
[Debug]   Starting builds...
[Debug]   pihole: Using platform option for build: linux/arm/v6
```
The build does not seem to be actually starting on the device and the command hangs there.

# Steps to Reproduce the Problem

Use a compose file defining a build context instead of referencing an external image:
```
version: "2.4"

services:
  pihole:
    build: pihole
```
The contents of the dockerfile in pihole folder does not seem to matter, the build fails as soon as you use a subfolder.

# Specifications

- **balena CLI version:** 17.4.9
- **Cloud backend: openBalena or balenaCloud?** balenaCloud
- **Operating system version:** linux debian vm on chromeos 119.0.6045.212
- **32/64 bit OS and processor:** Core i5 10021U
- **Install method:** zip package

**Comment by deanMike** [comment]
I've been experiencing the same issue recently. I've tried 17.4.9, 17.4.10, and 17.4.11 and it just hangs on all of them if you try to build a service rather than use an image.

**Comment by InvisibleSymbol** [comment]
I have been struggling with the same issue. Tried various things, none of which worked, which lead me here.

While testing various things I noticed that commands such as `balena device purge` start failing with `BalenaRequestError: Request error: Application with ID 2106642 is not in the current state` as soon as you enable local-mode, and start working again once you disable it. `balena logs` also hangs, even though new log entries can be observed when checking manually using `balena ssh`.

I was suspecting that this might be a bug in balena os instead of the cli but using os v4.0.26+rev1 with supervisor v14.13.4 didn't fix the issue (v5.0.8 & v15.2.0 was the default).
